File: python/p032.py
#!/usr/bin/python

from common import split_num

result = set()

for a in range(2, 100):
    b_start = 123 if a > 9 else 1234
    b_end = 10000 // a + 1
    for b in range(b_start, b_end):
        c = a * b
        nums = set(list(split_num(a)) + list(split_num(b)) + list(split_num(c)))
        if not nums ^ {1, 2, 3, 4, 5, 6, 7, 8, 9}:
            result.add(c)


print(sum(result))

File: rust/src/bin/p032.rs

use std::collections::HashSet;

mod common;
use common::split_num;

fn main() {

    let mut result: HashSet<usize> = HashSet::new();

    for a in (2..100) {
        let b_start = if a > 9 { 123 } else { 1234 };
        let b_end = 10000;
        for b in (b_start..b_end) {
            let c = a * b;
            let mut nums: Vec<_> = split_num(a).iter().chain(split_num(b).iter()).chain(split_num(c).iter()).cloned().collect();
            if nums.len() == 9 {
                nums.sort();
                if nums == [1, 2, 3, 4, 5, 6, 7, 8, 9] {
                    result.insert(c);
                }
            }
        }
    }

    println!("{}", result.iter().fold(0, |acc, x| acc + x));

}
